{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "42a14256-91c8-446e-92a7-ab6bf11055d3",
   "metadata": {},
   "source": [
    "# Python assignment 3\n",
    "\n",
    "## Topics to cover\n",
    "\n",
    "-   Working with files: Input/Output\n",
    "-   More with lists\n",
    "-   `sys.argv`, `re.search`\n",
    "-   `for` loops, processing files one line at a time.\n",
    "\n",
    "<p>\n",
    " \n",
    "</p>\n",
    "\n",
    "### This week you will work more with lists, and more importantly, will get comfortable reading IN and writing OUT. We are not literally reading in the contents of entire files at once, rather we are making connections (file objects or handles) between our program and these files. For each problem you will need to use a `for` loop to read data in one line at a time. The python_3\\_primer.md and slides from class should help to get you started.\n",
    "\n",
    "<p>\n",
    " \n",
    "</p>\n",
    "1).Read in the input file **genenames.txt** line by line, and add the\n",
    "values from the second “column” of each line to a list. This file has\n",
    "two columns, the second of which lists gene names. Print the list to\n",
    "check that your script has worked. The output should be the second\n",
    "“column” of the file genenames.txt.\n",
    "<p>\n",
    " \n",
    "</p>\n",
    "\n",
    "**Hints**: You might want to use two lists, the first of which might be\n",
    "created by splitting each line into a list so that you can extract the\n",
    "second element from it. 2. You may need to ‘add’ elements to the second\n",
    "list (using `.append`) with each iteration of a while loop.\n",
    "\n",
    "<p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "61800fa1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['carB', 'rpoD', 'rpoZ', 'mutL-1', 'sucA', 'ubiA', 'infB']\n"
     ]
    }
   ],
   "source": [
    "# Store the gene names in an empty list\n",
    "gene_names = []\n",
    "\n",
    "# Open the input file 'genenames.txt' for reading (assuming it's in the same directory as your notebook)\n",
    "with open('genenames.txt', 'r') as file:\n",
    "    # Read the file line by line\n",
    "    for line in file:\n",
    "        # Using the line.split to split each line into a list using a space as the delimiter\n",
    "        columns = line.split()\n",
    "        \n",
    "        # If the line contains at least two columns\n",
    "        if len(columns) >= 2:\n",
    "            # Extract and append the value from the second column to the gene_names list\n",
    "            gene_names.append(columns[1])\n",
    "\n",
    "# Print the list of gene names\n",
    "print(gene_names)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77a2ee8",
   "metadata": {},
   "source": [
    "2.) Now write a similar script to read the data in flies.txt. This file\n",
    "has three columns: fly number, day or night, and seconds of flying time\n",
    "recorded. You want to read in the seconds of flying time into an list,\n",
    "but ONLY for the night-time values. Write the information in this list\n",
    "to an outfile. This problem is similar to number 1, but is asking you\n",
    "add (maybe using the list function `.append`) a value onto a list if a\n",
    "different column in the file has a certain condition (N in this case).\n",
    "Instead of hardcoding the file name into your script, use `sys.argv` to\n",
    "open a file object from the command line. You may want to initialize\n",
    "your list before you get into the for loop.\n",
    "<p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65728ae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "# Check if the correct number of command-line arguments is provided\n",
    "if len(sys.argv) != 3:\n",
    "    print(\"Usage: python script.py input_file output_file\")\n",
    "    sys.exit(1)\n",
    "\n",
    "# Get the input and output file names from command-line arguments\n",
    "input_file = sys.argv[1]\n",
    "output_file = sys.argv[2]\n",
    "\n",
    "# Create an empty list to store the flying time for night-time values\n",
    "night_flying_time = []\n",
    "\n",
    "# Open the input file for reading\n",
    "with open(input_file, 'r') as infile:\n",
    "    # Open the output file for writing\n",
    "    with open(output_file, 'w') as outfile:\n",
    "        # Read the input file line by line\n",
    "        for line in infile:\n",
    "            # Split each line into a list using whitespace as the delimiter\n",
    "            columns = line.split()\n",
    "            \n",
    "            # Check if the second column is 'N' (night-time)\n",
    "            if len(columns) >= 2 and columns[1] == 'N':\n",
    "                # If it's night-time, append the flying time (third column) to the night_flying_time list\n",
    "                night_flying_time.append(columns[2])\n",
    "        \n",
    "        # Write the night flying times to the output file, one per line\n",
    "        outfile.write('\\n'.join(night_flying_time))\n",
    "\n",
    "print(\"Night-time flying times written to\", output_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0bd80de",
   "metadata": {},
   "source": [
    "#### Here is the output from the code above\n",
    "\n",
    "`202`<br>\n",
    "`300`<br>\n",
    "`167`<br>\n",
    "`301`<br>\n",
    "`180`<br>\n",
    "`202`<br>\n",
    "`300`<br>\n",
    "`167`<br>\n",
    "`301`<br>\n",
    "`180`<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99e8bb22",
   "metadata": {},
   "source": [
    "3.) Some practice working with DNA sequence data. Here we are going to\n",
    "process a fasta file (no60_intron_IME_data.fasta) containing DNA\n",
    "sequence data for 59,260 intron sequences You will need to calculate\n",
    "some summary information for each sequence, and to write that\n",
    "information to a file that you are going to create in your script.\n",
    "\n",
    "Instead of hardcoding the file name into your script, use `sys.argv` to\n",
    "open a file object from the command lineWrite to an output file:\n",
    "\n",
    "**A. the id lines**\n",
    "\n",
    "**B. the length of each sequence**\n",
    "\n",
    "**C. the GC content of each sequence (remember, GC content is simply the\n",
    "proportion of bases in a DNA sequence that are either G or C).**\n",
    "\n",
    "**Lastly have your program print to screen the number of DNA sequences\n",
    "(use incrementing, e.g., TotalSeq += 1, for each time you process a DNA\n",
    "sequence) in the no60_intron_IME_data.fasta file and the average GC\n",
    "content over all of the sequences.**\n",
    "\n",
    "**Hints:**\n",
    "\n",
    "1). Use a `for` loop to process the file one line at a time. You may\n",
    "want to remove the line endings, but then you can accomplish your goals\n",
    "by just working with each line as a string.\n",
    "\n",
    "2). If you remember, `str.count('G')`, for example, can be used to count\n",
    "the number of Gs in a string of DNA sequence data.\n",
    "\n",
    "3). `+=` will be useful, as you will want to keep a running total of a\n",
    "few variables each iteration through your for loop in order to get the\n",
    "total counts of C and G and the total equence length necessary for\n",
    "calculating the GC content for ALL sequences in the file together.\n",
    "\n",
    "4).Each ID line starts with a “\\>”. You will not want to perform any\n",
    "calculations on this line, but will want to print it to OUT. Hence you\n",
    "will need to recognize lines that start with “\\>”. A regular expression\n",
    "can do that for you. Maybe something like:\n",
    "\n",
    "    if re.search(\"^>\", DNAseq):\n",
    "\n",
    "For lines that start with A, T, C, or G, you will want to do something\n",
    "else.\n",
    "\n",
    "5).To calculate the GC content for teh entire collection of sequences,\n",
    "you will want to keep a running total of the counts of A, T, G, and C,\n",
    "as well as the total sequence length, and then perform a calculation on\n",
    "those running sums *outside of the `for` loop.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d81689f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import re\n",
    "\n",
    "# Get the input and output file names from command-line arguments\n",
    "input_file = sys.argv[1]\n",
    "output_file = sys.argv[2]\n",
    "\n",
    "# Initialize variables for final information\n",
    "total_sequences = 0\n",
    "total_length = 0\n",
    "total_GC_count = 0\n",
    "\n",
    "# Open the input file for reading\n",
    "with open(input_file, 'r') as infile:\n",
    "    # Open the output file for writing\n",
    "    with open(output_file, 'w') as outfile:\n",
    "        # Process the file line by line\n",
    "        for line in infile:\n",
    "            # Remove newline characters\n",
    "            line = line.strip()\n",
    "\n",
    "            # Check if the line starts with \">\"\n",
    "            if re.search(\"^>\", line):\n",
    "                # If it's an ID line, write it to the output file\n",
    "                outfile.write(line + '\\n')\n",
    "            else:\n",
    "                # If it's a sequence line, calculate and write the required information\n",
    "                sequence_length = len(line)\n",
    "                GC_count = line.count('G') + line.count('C')\n",
    "\n",
    "                # Write the length and GC content to the output file\n",
    "                outfile.write(f\"Length: {sequence_length}, GC content: {GC_count / sequence_length:.2%}\\n\")\n",
    "\n",
    "                # Update running totals\n",
    "                total_sequences += 1\n",
    "                total_length += sequence_length\n",
    "                total_GC_count += GC_count\n",
    "\n",
    "# Calculate average GC content\n",
    "average_GC_content = total_GC_count / total_length if total_length > 0 else 0\n",
    "\n",
    "# Print summary information to the screen\n",
    "print(\"Total DNA sequences:\", total_sequences)\n",
    "print(\"Average GC content:\", f\"{average_GC_content:.2%}\")\n",
    "\n",
    "# Append summary information to the output file\n",
    "with open(output_file, 'a') as outfile:\n",
    "    outfile.write(f\"\\nTotal DNA sequences: {total_sequences}\\n\")\n",
    "    outfile.write(f\"Average GC content: {average_GC_content:.2%}\\n\")\n",
    "\n",
    "print(\"Summary information written to\", output_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "014c8f37",
   "metadata": {},
   "source": [
    "`(base) jeshurunbiney@JESHURUNs-MacBook-Pro week8_python3 % python argv3.py no60_intron_IME_data.fasta new_GC_content.fasta`\n",
    "\n",
    "Total DNA sequences: 59260 <br>\n",
    "Average GC content: 32.36% <br>\n",
    "Summary information written to new_GC_content.fasta <br>\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
